libs/stm32f0_drivers/usart.c

```diff
@@ -0,0 +1,295 @@
+#include <stdio.h>
+#include <errno.h>
+
+#include <FreeRTOS.h>
+#include <timers.h>
+#include <task.h>
+#include <queue.h>
+#include <semphr.h>
+
+#include <stm32f0xx_usart.h>
+#include <stm32f0xx_misc.h>
+#include <stm32f0xx_gpio.h>
+#include <stm32f0xx_rcc.h>
+
+#include <usart.h>
+
+struct usart {
+	USART_TypeDef *periph;
+	bool autocrlf;
+	bool initialized;
+
+	struct {
+		xQueueHandle tx;
+		xQueueHandle rx;
+	} buffers;
+
+	struct {
+		void (*periph_clock_cmd)(uint32_t, FunctionalState);
+		uint32_t rccperiph;
+	} rcc;
+
+	struct {
+		GPIO_TypeDef *rxport;
+		uint16_t rxpin;
+		void (*rx_clock_cmd)(uint32_t, FunctionalState);
+		uint32_t rx_pin_clock;
+		uint8_t         rx_pinsource;
+		GPIO_TypeDef *txport;
+		uint16_t txpin;
+		void (*tx_clock_cmd)(uint32_t, FunctionalState);
+		uint32_t tx_pin_clock;
+		uint8_t         tx_pinsource;
+	} gpios;
+
+	bool rx_overflow;
+	uint32_t irq_no;
+} usarts[] = {
+#ifdef USART1_ENABLE
+	{
+		.periph = USART1,
+		.autocrlf = false,
+		.initialized = false,
+		.rcc = {
+			.periph_clock_cmd = RCC_APB2PeriphClockCmd,
+			.rccperiph = RCC_APB2Periph_USART1,
+		},
+		.gpios = {
+			.rxport = GPIOA,
+			.rxpin = GPIO_Pin_10,
+			.rx_clock_cmd = RCC_AHBPeriphClockCmd,
+			.rx_pin_clock = RCC_AHBPeriph_GPIOA,
+			.rx_pinsource = GPIO_PinSource10,
+			.txport = GPIOA,
+			.txpin = GPIO_Pin_9,
+			.tx_clock_cmd = RCC_AHBPeriphClockCmd,
+			.tx_pin_clock = RCC_AHBPeriph_GPIOA,
+			.tx_pinsource = GPIO_PinSource9,
+		},
+		.irq_no = USART1_IRQn,
+		.rx_overflow = false,
+	},
+#endif
+#ifdef USART2_ENABLE
+	{
+		.periph = USART2,
+		.autocrlf = false,
+		.initialized = false,
+		.rcc = {
+			.periph_clock_cmd = RCC_APB1PeriphClockCmd,
+			.rccperiph = RCC_APB1Periph_USART2,
+		},
+		.gpios = {
+			.rxport = GPIOA,
+			.rxpin = GPIO_Pin_3,
+			.rx_clock_cmd = RCC_APB2PeriphClockCmd,
+			.rx_pin_clock = RCC_APB2Periph_GPIOA,
+			.rx_pinsource = GPIO_PinSource3,
+			.txport = GPIOA,
+			.txpin = GPIO_Pin_2,
+			.tx_clock_cmd = RCC_APB2PeriphClockCmd,
+			.tx_pin_clock = RCC_APB2Periph_GPIOA,
+			.tx_pinsource = GPIO_PinSource2,
+		},
+		.irq_no = USART2_IRQn,
+		.rx_overflow = false,
+	},
+#endif
+};
+
+static struct usart *usart_lookup(uint8_t usart_no)
+{
+
```

Comment:
USART1 vs USART_1 and USART2 vs USART_2 can be a little dicey. I'd recommend to make the two definitions (periph vs usart_no) be a bit more different, to avoid "oopsie typo" moments.